File: BinarySearch.java
import java.io.*;
import java.util.*;

public class BinarySearch {

    static class Entry {
        long number;
        String text;

        Entry(long number, String text) {
            this.number = number;
            this.text = text;
        }
    }

    public static List<Entry> loadDataset(String filename) throws IOException {
        List<Entry> data = new ArrayList<>();
        BufferedReader reader = new BufferedReader(new FileReader(filename));
        String line;
        while ((line = reader.readLine()) != null) {
            String[] parts = line.split(",", 2);
            if (parts.length == 2) {
                long num = Long.parseLong(parts[0]);
                String text = parts[1];
                data.add(new Entry(num, text));
            }
        }
        reader.close();
        return data;
    }

    public static int binarySearch(List<Entry> data, long target) {
        int left = 0, right = data.size() - 1;
        while (left <= right) {
            int mid = (left + right) / 2;
            if (data.get(mid).number == target)
                return mid;
            else if (data.get(mid).number < target)
                left = mid + 1;
            else
                right = mid - 1;
        }
        return -1;
    }

    public static List<String> testBinarySearch(List<Entry> data) {
        int n = data.size();
        List<String> results = new ArrayList<>();

        long bestCase = data.get(n / 2).number;
        long avgCase = data.get(n / 4).number;
        long worstCase = 10_000_000_000L;

        results.add(measureCase("Best case", bestCase, data, n));
        results.add(measureCase("Average case", avgCase, data, n));
        results.add(measureCase("Worst case", worstCase, data, n));

        return results;
    }

    public static String measureCase(String caseName, long value, List<Entry> data, int repetitions) {
        long start = System.nanoTime();
        for (int i = 0; i < repetitions; i++) {
            binarySearch(data, value);
        }
        long end = System.nanoTime();
        double elapsedMs = (end - start) / 1_000_000.0;
        return caseName + " time: " + String.format("%.2f ms", elapsedMs);
    }

    public static void main(String[] args) {
        try (Scanner scanner = new Scanner(System.in)) {
            System.out.print("Enter sorted dataset filename: ");
            String filename = scanner.nextLine().trim();

            File file = new File(filename);
            if (!file.exists()) {
                System.out.println("Error: File not found.");
                return;
            }

            List<Entry> dataset = loadDataset(filename);
            List<String> results = testBinarySearch(dataset);

            
            String baseName = file.getName();
            String[] nameParts = baseName.split("_");
            String sizePart = nameParts.length > 1 ? nameParts[nameParts.length - 1].replace(".csv", "") : "result";
            String outputFile = "binary_search_" + sizePart + ".txt";

            BufferedWriter writer = new BufferedWriter(new FileWriter(outputFile));
            for (String line : results) {
                writer.write(line);
                writer.newLine();
            }
            writer.close();

            System.out.println("\nResults written to " + outputFile);
            for (String line : results) {
                System.out.println(line);
            }

        } catch (IOException e) {
            System.out.println("Error handling files.");
        } catch (NumberFormatException e) {
            System.out.println("Error: Invalid number format.");
        }
    }
}

File: binary_search.py
import csv
import time
import os

def binary_search(arr, target):
    left, right = 0, len(arr) - 1
    while left <= right:
        mid = (left + right) // 2
        if arr[mid][0] == target:
            return mid
        elif arr[mid][0] < target:
            left = mid + 1
        else:
            right = mid - 1
    return -1

def load_dataset(filename):
    with open(filename, 'r') as csvfile:
        reader = csv.reader(csvfile)
        data = [(int(row[0]), row[1]) for row in reader]
    return data

def test_binary_search(data):
    n = len(data)
    results = []

    # Choose values to test
    best_case = data[n // 2][0]
    average_case = data[n // 4][0]
    worst_case = 10**10  

    def measure_case(case_name, value):
        start = time.time()
        for _ in range(n):
            binary_search(data, value)
        end = time.time()
        elapsed = (end - start) * 1000  
        results.append(f"{case_name} time: {elapsed:.2f} ms")

    measure_case("Best case", best_case)
    measure_case("Average case", average_case)
    measure_case("Worst case", worst_case)

    return results


if __name__ == "__main__":
    filename = input("Enter sorted dataset filename : ").strip()

    if not os.path.exists(filename):
        print("Error: File not found.")
        exit(1)

    data = load_dataset(filename)
    results = test_binary_search(data)

    
    base_name = os.path.basename(filename)
    size_part = base_name.split("_")[-1].split(".")[0]
    output_file = f"binary_search_{size_part}.txt"

    with open(output_file, "w") as f:
        for line in results:
            f.write(line + "\n")

    print(f"\nResults written to {output_file}")
    print("\n".join(results))
